clients/src/main/java/org/apache/kafka/common/record/MemoryRecords.java

```diff
@@ -232,28 +271,16 @@ public RecordsIterator(ByteBuffer buffer, CompressionType type, boolean shallow)
         protected LogEntry makeNext() {
             if (innerDone()) {
                 try {
-                    // read the offset
-                    long offset = stream.readLong();
-                    // read record size
-                    int size = stream.readInt();
-                    if (size < 0)
-                        throw new IllegalStateException("Record with size " + size);
-                    // read the record, if compression is used we cannot depend on size
-                    // and hence has to do extra copy
-                    ByteBuffer rec;
-                    if (type == CompressionType.NONE) {
-                        rec = buffer.slice();
-                        int newPos = buffer.position() + size;
-                        if (newPos > buffer.limit())
-                            return allDone();
-                        buffer.position(newPos);
-                        rec.limit(size);
-                    } else {
-                        byte[] recordBuffer = new byte[size];
-                        stream.readFully(recordBuffer, 0, size);
-                        rec = ByteBuffer.wrap(recordBuffer);
+                    LogEntry entry = getNextEntry();
+                    // No more record to return.
+                    if (entry == null)
+                        return allDone();
+
+                    // Convert offset to absolute offset if needed.
+                    if (absoluteBaseOffset >= 0) {
```

Comment:
Hmm, absoluteBaseOffset is only non-negative for inner iterators. The outer iterators always have absoluteBaseOffset=-1. Because we create a separate independent inner iterator for each compressed message set, the same absoluteBaseOffset should be used by all the messages in that compressed message set. After we finish iterating one compressed message set and return to the outer iterator, the absoluteBaseOffset of inner iterator will not affect the outer iterator, i.e. the outer iterator absoluteBaseOffset remains -1. So even the next record is an uncompressed record, it should not be affected.